### PR TITLE
[tt-train] Split TT-Train L2 perf tests into per-model matrix jobs and add more tests

### PR DESCRIPTION
**Pull request by kevinwuTT**
### Summary

Split the TT-Train L2 perf job into per-model matrix entries so each model runs as its own CI job, with pretraining and finetuning tests scheduled as separate L2 nightly workflows. Also add GRPO examples and C++ models.

### Changes
- Change one `run_models.py` job to individual model jobs with per-SKU timeouts and `pretraining` / `finetuning` categories
- Split into `Run TT-Train Pretraining Tests` and `Run TT-Train Finetuning Tests` with separate dispatch inputs
- Added GRPO examples under `tt-train/sources/examples/grpo/`
- Readded NanoGPT, NanoLlama, TInyLlama C++ examples
- Set `HF_TOKEN` for Llama 3.2 1B
- Moved model config to `run_models_configs/single_cards.yaml`
- Updated memory logging in C++ `nano_gpt` and parsing in `analyze_memory.py`

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ctr-kevinwu/split-models-l2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ctr-kevinwu/split-models-l2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:ctr-kevinwu/split-models-l2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ctr-kevinwu/split-models-l2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ctr-kevinwu/split-models-l2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ctr-kevinwu/split-models-l2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ctr-kevinwu/split-models-l2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ctr-kevinwu/split-models-l2)

### CODE REVIEW

File: .github/workflows/tt-train-tests.yaml [hunk 1]
```diff
@@ -89,6 +94,17 @@ jobs:
             "${{ inputs.enabled-skus }}" \
             ./.github/sku_config.yaml
 
+      - name: Filter matrix by subcategories
+        id: filter-categories
+        run: |
+          matrix='${{ steps.build-matrix.outputs.matrix }}'
+          categories='${{ inputs.subcategories }}'
+          filtered=$(echo "$matrix" | jq -c --arg cats ",$categories," '[.[] | .category as $c | select($cats | contains(",\($c),"))]')
```
[the remark this answers is not in the document]
Comment (kevinwuTT) [reply .github/workflows/tt-train-tests.yaml:103]: Added job conditionals to the step https://github.com/tenstorrent/tt-metal/pull/47830/changes/e86366818912949f80efd4b2e9505c83478419ee

**Review by tdowdallTT — APPROVED** [review]

**Comment by ichovpanTT** [comment]
Just interested, how much time does a boolq training and boolq accuracy run takes in CI? Is it on 1 card or on multiple?

**Comment by kevinwuTT** [comment]
> Just interested, how much time does a boolq training and boolq accuracy run takes in CI? Is it on 1 card or on multiple?

@ichovpanTT Looks like 40 minutes for BoolQ training and about 3-4 minutes for BoolQ Accuracy on a single P150.

https://github.com/tenstorrent/tt-metal/actions/runs/28107425290/job/83262485600

And this is running the script without any extra parameters, i.e.
`python tt-train/sources/examples/grpo/boolq_training_example.py`
`python tt-train/sources/examples/grpo/boolq_accuracy_example.py`

**Review by philei-tt — APPROVED** [review]
